### DISCUSSION TITLE
Improve concurrency when synchronizing

### DISCUSSION THREAD

**Discussion by gentledepp** [Mimetis/Dotmim.Sync#696] [Ideas]
- dotmim-sync does use a single transaction to read all the changes of one scope.
- Say your scope is spanning dozens of tables, then there are quite a bunch of read-locks on them until all changes have been read.

-> Now if another user syncs then applying his changes may fail due to a transaction deadlock.

We solved this by creating a separate version of dotmim.sync that uses one transaction per table when fetching changes.
This does have the downside, that you _may_ see records that you were not supposed to see. 
eg. `Order`, `OrderDetail` tables. 
`OrderDetail` has foreign key to `Order`. Therefore you have to synchronzie these tables in the order 
- `Order`
- `OrderDetail`
Now using our approach you could end up with an `OrderDetail` record sent to your client, that points to an `Order` that was not synchronized to your device yet.

The best option, in this case (and in my humble opinion), would be to take the "currentTimestamp" (taken the first instant the sync request receives the server) and, instead of using transactions for selecting changes, create a separate transaction for each table and 
`select all changes where timestamp > lastSyncedTimestamp && timestamp < currentTimestamp` 

That would give you the best of both worlds:
- no single transaction
- no non-consistent change sets

@Mimetis what do you think?

**Comment by Mimetis** [comment]
## Transaction

What's the benefit to have one transaction per table (for reading) ?
I mean, if we reduce the transaction to a single table (for reading), there is ... basically no need to create a transaction ?

That's being said, should we consider to be able to remove transactions when reading changes  (at least through an option or a mechanism like interceptors ?)

Regarding the issue #695, should we also consider to be able to remove transactions for writing as well ?

For the read lock, you have the option to set an `IsolationLevel` on your provider.
Is it enough ?

https://github.com/Mimetis/Dotmim.Sync/blob/f992c9e11239bd39afa375d3aab6e4410ce16d8d/Projects/Dotmim.Sync.Core/CoreProvider.cs#L88

I don't know if removing transactions will have more benefits than drawbacks, but adding an option to remove transaction during a particular step should not be too complicated, I guess.
But let's discuss it to be sure I understand everything :)

## Timestamp

For the timestamp, in a nutshell, this value does not have any relationship with the in progress transaction.
- `lastSyncedTimestamp` is stored (copying the `currentTimestamp` value once the sync is done)
- `currentTimestamp` is generated just before making the sync process (just before getting changes) on each side 
  - Doing this prevents to have "ghosts" records that could be inserted / updated during the selecting changes step.
  - If we have a record inserted / updated during the getting changes step, well this record will be synced on the next sync.

https://github.com/Mimetis/Dotmim.Sync/blob/f992c9e11239bd39afa375d3aab6e4410ce16d8d/Projects/Dotmim.Sync.Core/Orchestrators/LocalOrchestrator.cs#L112

https://github.com/Mimetis/Dotmim.Sync/blob/f992c9e11239bd39afa375d3aab6e4410ce16d8d/Projects/Dotmim.Sync.Core/Orchestrators/RemoteOrchestrator.cs#L355

There is no relationships between transactions and timestamp, because (especially on SQL Server) we are not using the `min_active_rowversion` version (based on transaction) because of the effect you are describing

https://github.com/Mimetis/Dotmim.Sync/blob/f992c9e11239bd39afa375d3aab6e4410ce16d8d/Projects/Dotmim.Sync.SqlServer/Builders/SqlScopeBuilder.cs#L223

**Reply by gentledepp** [reply]
# Transaction
You are right - using a transaction per table is like using no transaction at all.
And that is exactly the point.
The main difference here is, that we explicitly create a transaction for each sp-select call instead of just firing the sql-select on its own (which would cause an implicit transaction being created by MS SQL Server)
To be honest: The reason we create the transactions explicitly is only so that we had to change as little code as possible in dotmim.sync


# Timestamp
I am aware of how timestamps are working in Dotmim.Sync.
Maybe I did not explain this clearly enough:

I want to avoid that in a scope of tables where records depend on each other (orderdetail depends on order because of its foreign key) some "child" record is synced to the client, while the "parent" record is not. That would cause the sync to fail (at least if the client database did not disable checkling FK constraints)

The idea to solve this _potential_ issue using timestamps.

1. client sends a getchanges request providing the `LastSyncedTimestamp` which is then used as `@sync_min_timestamp`
2. the server (using sql server) gets that request
3. ... and first retrieves the `MIN_ACTIVE_ROWVERSION` that I'll call now `@CurrentTimeStamp`
4. then we can (query all the changes across tables where `[side].[timestamp] > @sync_min_timestamp AND [side].[timestamp] < @CurrentTimeStmap`
This would allow us to
- get a consistent set of records of a given scope 
- without having the need to ensure this using transactions

Does it make more sense now?
(@Mimetis  thanks for the quick response, but it is not urgent - pleaes take your time ;-))

**Reply by Mimetis** [reply]
> ... and first retrieves the MIN_ACTIVE_ROWVERSION that I'll call now @CurrentTimeStamp
>then we can (query all the changes across tables where [side].[timestamp] > @sync_min_timestamp AND [side].[timestamp] < @CurrentTimeStmap

This is exactly how it's working :)

`CurrentTimeStamp` is retrieved before getting changes (and becomes later `LastSyncedTimestamp` )

Did you see the code I pointed ? Or do I misunderstand something ?

**Reply by gentledepp** [reply]
No, I do not see the code. 
But `CurrentTimeStamp` is retrieved by the **client**.
What I mean is using the `MIN_ACTIVE_ROWVERSION` on the server-side to actually **limit** the records to not be **too new** (`[side].[timestamp] < @CurrentTimeStamp`)

Dis was for sure not the case in the 'old' DMS framework. Did you actually change this behavior?
If so, why do you still need to have a transaction spanning all calls to sp_selectchanges for one complete scope?

**Reply by Mimetis** [reply]
`CurrentTimeStamp` is stored on the client (that way, there is no dependency to the client on the server) but is retrieved on the server, just before making a getting changes (Mainly because server timestamp can be completely different from the client (ie SQL Server uses a Timestamp meanwhile SQLite is using a datetime)

What happens:
- Client sends changes to be applied on server, with `LastTimeStamp`.
- Applying changes from client to the server in a transaction.
- Getting server local timestamp `CurrentTimeStamp`.
- Getting changes from `LastTimeStamp` to `CurrentTimeStamp`.
- Send back changes and `CurrentTimeStamp` to client.
- Client stores this `CurrentTimeStamp` as `LastTimeStamp`.

I think that behaviors has changed since the old `DMS` Fx.
And indeed, not sure we need a transaction for all get changes calls

**Reply by gentledepp** [reply]
nice! You do not need a transaction then because you already limit the record set so you do not get changes that were committed before the client started synchronizing 👍
This is the only way to keep the sync set consistent.
Because even when having a single read-committed transaction spanning all select-changes calls, it could happen that
1) you get changes from the `Orders` table
2) the other transaction commits a new order `o` and some order details  `od1` and `od2`
3) you now get the (read-committed) changes from the `OrderDetails` table
-> you'd now include `od1` and `od2` in your changeset which would then fail to be inserted on the client, because `o` is not included in the changeset and the order details refer to it.

Baseline: The single transaction does not really do anything well. It raises the possibility of sql-deadlocks though (which we had to fight a lot in the _old DMS FX_

**Reply by Mimetis** [reply]
let me rephrase your scenario, I think we don't get the `OrderDetails` :)

1. I get the `currentTimestamp` (after applying changes, and before getting any changes from any table)
2. I get changes from **Orders** table (from `lastTimeStamp` to `currentTimestamp`)
3. Another transaction commits a new order in **Orders** `o`and two **OrderDetails** `od1` & `od2`
   - `o` gets a  tracking timestamp value of `currentTimestamp +1`
   - `od1` gets a tracking timestamp value of `currentTimestamp +2`
   - `od3` gets a tracking timestamp value of `currentTimestamp +3`
4. I get changes from **OrderDetails** (from `lastTimeStamp` to `currentTimestamp`)
   - I don't get neither `od1`, neither `od2`
5. After getting all changes `lastTimestamp` becomes `currentTimestamp`
6. On the next sync, I will get `o`, `od1` and `od2`

Make sense ?

**Reply by gentledepp** [reply]
yes - this is **perfect** indeed.
This way, you can also safely get rid of the "single transaction spanning all sp_select calls" as it brings  no benefits but has the downside of deadlocks (if other clients try to write to a table that has a table-lock because the changeset is huge - eg.)

**Reply by Mimetis** [reply]
I'm on it ;)

**Reply by gentledepp** [reply]
take your time. This is only a *discussion* that I merely started to not forget this topic ;-)
